webserver/cors/src/main/java/io/helidon/webserver/cors/LogHelper.java

```diff
@@ -0,0 +1,133 @@
+
+    /**
+     * Collects headers for assignment to a request or response and logging during assignment.
+     */
+    static class Headers {
+        private final List<Map.Entry<String, Object>> headers = new ArrayList<>();
+        private final List<String> notes = LOGGER.isLoggable(DECISION_LEVEL) ? new ArrayList<>() : null;
+
+        Headers add(String key, Object value) {
+            headers.add(new AbstractMap.SimpleEntry<>(key, value));
+            return this;
+        }
+
+        Headers add(String key, Object value, String note) {
+            add(key, value);
+            if (notes != null) {
+                notes.add(note);
+            }
+            return this;
+        }
+
+        void set(BiConsumer<String, Object> consumer, String note) {
+            headers.forEach(entry -> consumer.accept(entry.getKey(), entry.getValue()));
+            LOGGER.log(DECISION_LEVEL, () -> note + ": " + headers + (notes == null ? "" : notes));
+        }
+    }
+
+    static <T> boolean isRequestTypeNormal(boolean result, RequestAdapter<T> requestAdapter, Optional<String> originOpt,
+            Optional<String> hostOpt) {
+        // If no origin header or same as host, then just normal
+
+        List<String> reasonsWhyNormal = new ArrayList<>();
+        List<String> factorsWhyCrossHost = new ArrayList<>();
+
+        if (originOpt.isEmpty()) {
+            reasonsWhyNormal.add("header " + ORIGIN + " is absent");
+        } else {
+            factorsWhyCrossHost.add(String.format("header %s is present (%s)", ORIGIN, originOpt.get()));
+        }
+
+        if (hostOpt.isEmpty()) {
+            reasonsWhyNormal.add("header " + HOST + " is absent");
+        } else {
+            factorsWhyCrossHost.add(String.format("header %s is present (%s)", HOST, hostOpt.get()));
+        }
+
+        if (hostOpt.isPresent() && originOpt.isPresent()) {
+            String partOfOriginMatchingHost = "://" + hostOpt.get();
+            if (originOpt.get()
+                    .contains(partOfOriginMatchingHost)) {
+                reasonsWhyNormal.add(String.format("header %s '%s' matches header %s '%s'; not cross-host", ORIGIN,
+                        originOpt.get(), HOST, hostOpt.get()));
+            } else {
+                factorsWhyCrossHost.add(String.format("header %s (%s) does not match header %s %s; cross-host", ORIGIN,
+                        originOpt.get(), HOST, hostOpt.get()));
+            }
+        }
+
+        if (result) {
+            LOGGER.log(LogHelper.DECISION_LEVEL,
+                    () -> String.format("Request %s is not cross-host: %s", requestAdapter, reasonsWhyNormal));
+        } else {
+            LOGGER.log(LogHelper.DECISION_LEVEL,
+                    () -> String.format("Request %s is cross-host: %s", requestAdapter, factorsWhyCrossHost));
+        }
+        return result;
+    }
+
+    static <T> RequestType inferCORSRequestType(RequestType result, RequestAdapter<T> requestAdapter, String methodName,
```

Comment:
same as with previous method - `void` return type and name it with log in the name.
Also if this is all about logging, maybe checking that the log level is relevant would be good, before doing all the work (same for all methods in this class), as the first thing in every method:
```
if (!LOGGER.isLoggable(DECISION_LEVEL)) {
     return;
}
```

Reply:
For `isRequestTypeNormal` and `inferCORSRequestType` the calling code had already checked whether the logger was enabled at a fine enough level (and also checked a separate boolean as well). But to be more obvious I've revised both to pass the other boolean as an additional argument and then do the check inside the two methods.

(I had originally planned to use the logging methods to compute the values and (if enabled) do the logging. I chose to compute the value first in the caller for efficiency and then do the logging separately. I have set the return types to `void` and renamed the methods.)